xds/googledirectpath/googlec2p.go

```diff
@@ -0,0 +1,175 @@
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/google"
+	"google.golang.org/grpc/grpclog"
+	"google.golang.org/grpc/internal/googlecloud"
+	internalgrpclog "google.golang.org/grpc/internal/grpclog"
+	"google.golang.org/grpc/resolver"
+	xdsclient "google.golang.org/grpc/xds/internal/client"
+	"google.golang.org/grpc/xds/internal/client/bootstrap"
+	"google.golang.org/grpc/xds/internal/env"
+	"google.golang.org/grpc/xds/internal/version"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/structpb"
+
+	_ "google.golang.org/grpc/xds" // To register xds resolvers and balancers.
+)
+
+const (
+	c2pScheme = "google-c2p"
+
+	tdURL          = "directpath-trafficdirector.googleapis.com"
+	httpReqTimeout = 10 * time.Second
+	zoneURL        = "http://metadata.google.internal/computeMetadata/v1/instance/zone"
+	ipv6URL        = "http://metadata.google.internal/computeMetadata/v1/instance/network-interfaces/0/ipv6s"
+
+	gRPCUserAgentName               = "gRPC Go"
+	clientFeatureNoOverprovisioning = "envoy.lb.does_not_support_overprovisioning"
+	ipv6CapableMetadataName         = "TRAFFICDIRECTOR_DIRECTPATH_C2P_IPV6_CAPABLE"
+
+	logPrefix = "[google-c2p-resolver]"
+
+	dnsName, xdsName = "dns", "xds"
+)
+
+type xdsClientInterface interface {
+	Close()
+}
+
+// For overriding in unittests.
+var (
+	onGCE = googlecloud.OnGCE
+
+	newClientWithConfig = func(config *bootstrap.Config) (xdsClientInterface, error) {
+		c, err := xdsclient.NewWithConfig(config)
+		return c, err
+	}
+
+	logger = internalgrpclog.NewPrefixLogger(grpclog.Component("directpath"), logPrefix)
+
+	defaultNode = &v3corepb.Node{
+		Id:                   "C2P",
+		Metadata:             nil, // To be set if ipv6 is enabled.
+		Locality:             nil, // To be set to the value from metadata.
+		UserAgentName:        gRPCUserAgentName,
+		UserAgentVersionType: &v3corepb.Node_UserAgentVersion{UserAgentVersion: grpc.Version},
+		ClientFeatures:       []string{clientFeatureNoOverprovisioning},
+	}
+
+	ipv6EnabledMetadata = &structpb.Struct{
+		Fields: map[string]*structpb.Value{
+			ipv6CapableMetadataName: {
+				Kind: &structpb.Value_BoolValue{BoolValue: true},
+			},
+		},
+	}
+)
+
+func init() {
+	// TODO: remove this env var check when c2p resolver is proven stable.
+	if env.C2PResolverSupport {
+		resolver.Register(c2pResolverBuilder{})
+	}
+}
+
+type c2pResolverBuilder struct{}
+
+func (c2pResolverBuilder) Build(t resolver.Target, cc resolver.ClientConn, opts resolver.BuildOptions) (resolver.Resolver, error) {
+	if !runDirectPath() {
+		// If not xDS, fallback to DNS.
+		t.Scheme = dnsName
+		return resolver.Get(dnsName).Build(t, cc, opts)
+	}
+
+	// Note that the following calls to getZone() and getIPv6Capable() does I/O,
+	// and has 10 seconds timeout.
+	//
+	// This should be fine in most of the cases. In certain error cases, this
+	// could block Dial() for up to 20 seconds.
+	nodeCopy := proto.Clone(defaultNode).(*v3corepb.Node)
```

Comment:
Is it more efficient/simpler to make `defaultNode` a function that returns a new `Node` each time?  (It is better for type safety at least.)

Reply:
Done